### PR TITLE
Handle bad_token by removing bad refresh token from cache

### PR DESCRIPTION
**Pull request by hectormmg**
This PR:
- Catches `bad_token` sub error and handles it by immediately removing the refresh token used in the refresh request, then lets the error bubble up to be handled as it is today
- Adds test for this scenario

### CODE REVIEW

File: lib/msal-common/src/response/ResponseHandler.ts [hunk 1]
```diff
@@ -234,6 +238,23 @@ export class ResponseHandler {
                 );
             }
 
+            if (
```
Comment (hectormmg) [inline lib/msal-common/src/response/ResponseHandler.ts]: There's some duplicate logic here I'll most likely refactor this before merging.

File: lib/msal-common/src/error/InvalidGrantAuthError.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        desc: InvalidGrantAuthErrorMessages[
+            InvalidGrantAuthErrorCodes.badToken
+        ],
+    },
+};
+
+/**
+ * Error thrown when invalid grant is used.
+ */
+export class InvalidGrantAuthError extends AuthError {
```
Comment (tnorling) [inline lib/msal-common/src/error/InvalidGrantAuthError.ts]: Does this really need to be its own class? Why can't we continue to use InteractionRequiredAuthError/ServerError?

File: lib/msal-common/src/error/InvalidGrantAuthError.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+/**
+ * Helper function used to determine if an error thrown by the server requires bad token removal
+ * @param errorCode
+ * @param errorString
+ * @param subError
+ */
+export function isInvalidGrantAuthError(
```
Comment (tnorling) [inline lib/msal-common/src/error/InvalidGrantAuthError.ts]: I think the logic is simply `code === "invalid_grant"`

File: lib/msal-common/src/error/InvalidGrantAuthError.ts (hunk 2 above)
Comment (hectormmg) [reply lib/msal-common/src/error/InvalidGrantAuthError.ts]: Moved it back to InteractionRequired and added an exception for it in acquireTokenSilent's logic to determine if an error is silently resolvable

File: lib/msal-common/src/error/InteractionRequiredAuthError.ts [hunk 4]
```diff
@@ -32,6 +34,8 @@ const InteractionRequiredAuthErrorMessages = {
         "The requested account is not available in the native broker. It may have been deleted or logged out. Please sign-in again using an interactive API.",
     [InteractionRequiredAuthErrorCodes.refreshTokenExpired]:
         "Refresh token has expired.",
+    [InteractionRequiredAuthErrorCodes.badToken]:
+        "Server returned invalid_grant because of an unusable refresh token was used. Refresh token must be removed from cache.",
```
Comment (tnorling) [inline lib/msal-common/src/error/InteractionRequiredAuthError.ts]: Cache clearing is performed by MSAL so this bit doesn't need to be suggested.

```suggestion
        "Identity provider returned bad_token due to an expired or invalid refresh token. Please invoke an interactive API to resolve.",
```

File: lib/msal-browser/src/controllers/StandardController.ts [hunk 5]
```diff
@@ -2081,7 +2081,9 @@ export class StandardController implements IController {
                             const isSilentlyResolvable =
                                 (!(
                                     refreshTokenError instanceof
-                                    InteractionRequiredAuthError
+                                        InteractionRequiredAuthError &&
+                                    refreshTokenError.subError !==
```
Comment (tnorling) [inline lib/msal-browser/src/controllers/StandardController.ts]: This logic is really hard to follow. It may be time to factor this into a helper function with each condition separated to make it more readable with comments and such

**Review by tnorling — APPROVED** [review]
A couple minor non-blocking suggestions. Looks good otherwise.

**Review by peterzenz — APPROVED** [review]
